CODING_GUIDELINES.md

```diff
@@ -0,0 +1,64 @@
+# Coding Guidelines
+
+This document is an extension to [CONTRIBUTING](./CONTRIBUTING.md) and provides more details about the coding guidelines and requirements.
+
+## API & Design
+
+- Code must be well structured:
+  - packages must have a limited responsibility (different concerns can go to different packages),
+  - types must be easy to compose,
+  - think about maintainbility and testability.
+- "Depend upon abstractions, [not] concretions".
+- Try to limit the number of methods you are exposing. It's easier to expose something later than to hide it.
+- Take advantage of `internal` package concept.
+- Follow agreed-upon design patterns and naming conventions.
+- publicly-exposed functions are named logically, have forward-thinking arguments and return types.
+- Avoid global variables and global configurators.
+- Favor composable and extensible designs.
+- Minimize code duplication.
+- Limit third-party dependencies.
+
+Performance:
+
+- Avoid unnecessary operations or memory allocations.
+
+Security:
+
+- Pay proper attention to exploits involving:
+  - gas usage
+  - transaction verification and signatures
+  - malleability
+  - code must be always deterministic
+- Thread safety. If some functionality is not thread-safe, or uses something that is not thread-safe, then clearly indicate the risk on each level.
```

Comment:
here as well I would add some examples of scenarios where a code could be not thread-safe.

Reply:
Added example